### PR TITLE
Added Documentation for SDM class

### PR DESCRIPTION
**Pull request by ks147**
#### Brief description of what is fixed or changed
Added Documentation for SDM class

#### Release Notes

* polys
  * Added Documentation for SDM class

### CODE REVIEW

File: sympy/polys/matrices/sdm.py [hunk 1]
```diff
@@ -19,6 +19,33 @@ class SDM(dict):
...
+
+        Examples
+        ========
+
+        >>> from sympy.polys.matrices.sdm import SDM
+        >>> from sympy import QQ
+        >>> elemsdict = {0:{1:QQ(2)}, 1:{}}
+        >>> A = SDM(elemsdict, (2, 2), QQ)
+        >>> A
+        {0: {1: 2}, 1: {}}
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:46]: The `elemsdict` should not have any empty dicts in there. This docstring should explain that it is the responsibility of whoever creates `elemsdict` to remove zero rows and zero elements.

Also this should show how to construct a `DomainMatrix` from the `SDM`.

File: sympy/polys/matrices/sdm.py [hunk 2]
```diff
@@ -30,14 +57,82 @@ def __init__(self, elemsdict, shape, domain):
...
+        =======
+
+        :py:class:`~.SDM` containing elements of ddm
+
+        Examples
+        ========
+
+        >>> from sympy.polys.matrices.sdm import SDM
+        >>> from sympy import QQ
+        >>> ddm = [[QQ(1, 2), 0], [0, QQ(3, 4)]]
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:129]: The `0` should be `QQ(0)`.

File: sympy/polys/matrices/sdm.py (hunk 1 above)
Comment (oscarbenjamin) [reply sympy/polys/matrices/sdm.py:46]: I think it would be better to move this docstring to be the main class docstring and also to show examples of using the `SDM` e.g. for adding and multiplying

File: sympy/polys/matrices/sdm.py [hunk 3]
```diff
@@ -116,46 +324,202 @@ def matmul(A, B):
...
+        """
+
+        Returns LU decomposition for a matrix A
+
+        Examples
+        ========
+
+        >>> from sympy import QQ
+        >>> from sympy.polys.matrices.sdm import SDM
+        >>> A = SDM({0:{0:QQ(1), 1:QQ(2)}, 1:{0: QQ(3), 1: QQ(4)}}, (2, 2), QQ)
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:482]: The spacing around colons is inconsistent.

File: sympy/polys/matrices/sdm.py [hunk 4]
```diff
@@ -192,6 +570,22 @@ def hstack(A, *B):
...
+        The domain of the elements will be same as domain of the :py:class:`~.SDM`.
+
+        Examples
+        ========
+
+        >>> from sympy import QQ
+        >>> from sympy.polys.matrices.sdm import SDM
+        >>> A = SDM({0:{0:QQ(1), 1:QQ(2)}, 1:{0: QQ(3), 1: QQ(4)}}, (2, 2), QQ)
+        >>> A.charpoly()
+        [1, -5, -2]
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:585]: It would be good to show here how to construct a Poly from these e.g.:
```
x = Symbol('x')
p = Poly(A.charpoly(), x, domain=A.domain)
p
```

File: sympy/polys/matrices/sdm.py [hunk 5]
```diff
@@ -95,6 +278,31 @@ def __rmul__(a, b):
...
+
+        Examples
+        ========
+
+        >>> from sympy import ZZ
+        >>> from sympy.polys.matrices.sdm import SDM
+        >>> A = SDM({0:{1: ZZ(2)}, 1:{0:ZZ(1)}}, (2, 2), ZZ)
+        >>> B = SDM({0:{0:ZZ(2), 1:ZZ(3)}, 1:{0:ZZ(4)}}, (2, 2), ZZ)
+        >>> A.matmul(B)
+        {0: {0: 8}, 1: {0: 2, 1: 3}}
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:303]: For each of these methods it would be good to point out that they are also used by the operators so e.g. this could show that you get the same from `A*B`.

Comment (oscarbenjamin) [reply sympy/polys/matrices/sdm.py:303]: These docstrings should also explain when the domains are are required to match or what constraints there are on the shapes of the matrices for the operation to be defined.

File: sympy/polys/matrices/sdm.py (hunk 1 above)
Comment (ks147) [reply sympy/polys/matrices/sdm.py:46]: I'll do that

File: sympy/polys/matrices/sdm.py (hunk 2 above)
Comment (ks147) [reply sympy/polys/matrices/sdm.py:129]: I'll change it everywhere

File: sympy/polys/matrices/sdm.py (hunk 5 above)
Comment (ks147) [reply sympy/polys/matrices/sdm.py:303]: You actually don't get the same from `A * B`. 

>>> A * B
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for *: 'SDM' and 'SDM'

This is what you get with A*B

Comment (ks147) [reply sympy/polys/matrices/sdm.py:303]: I think ` __add__(A, B)` needs to be implemented for SDM class

Comment (oscarbenjamin) [reply sympy/polys/matrices/sdm.py:320]: Oh yeah, I think it's `A @ B` or `__matmul__`. Is that not implemented for `SDM`?

Comment (oscarbenjamin) [reply sympy/polys/matrices/sdm.py:320]: Perhaps it is better to use `*` anyway. Either way `SDM` and `DDM` should be made consistent so can you compare how `DDM` works?

Comment (ks147) [reply sympy/polys/matrices/sdm.py:320]: Consistent in what ways? You want both to have the same functionality right, is that what you're saying

Comment (ks147) [reply sympy/polys/matrices/sdm.py:320]: By the way `*` does not work in DDM either
```python
A = DDM([[ZZ(1), ZZ(2)], [ZZ(2), ZZ(3)]], (2, 2), ZZ)
B = DDM([[ZZ(3), ZZ(4)], [ZZ(1), ZZ(2)]], (2, 2), ZZ)
A * B
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for *: 'DDM' and 'DDM
```

Comment (oscarbenjamin) [reply sympy/polys/matrices/sdm.py:320]: I mean that when using methods like `A * B` or `A.matmul(B)` etc then all behaviour should be exactly the same whether `A` and `B` are both `DDM` or both `SDM`. These two classes are supposed to be completely interchangeable in that sense. So if you are changing any part of the behaviour of `SDM` you should check that `DDM` behaves in exactly the same way.

Comment (oscarbenjamin) [reply sympy/polys/matrices/sdm.py:336]: > I'll have to make `__mul__ ` in DDM consistent with definition in DomainMatrix

Note that `DDM` and `DomainMatrix` do not need to be consistent as they are not intended to be interchangeable. It is `DDM` and `SDM` that need to be consistent. For now I think that `SDM` should just use `@` for matrix multiplication just like `DDM` does. We can consider changing that in future but these two classes need to remain equivalent.

File: sympy/polys/matrices/sdm.py [hunk 6]
```diff
@@ -17,6 +17,43 @@ class SDM(dict):
...
+
+    We declare a 2x2 :py:class:`~.SDM` matrix belonging
+    to QQ domain as shown below.
+
+    >>> from sympy.polys.matrices.sdm import SDM
+    >>> from sympy import QQ
+    >>> elemsdict = {0:{1:QQ(2)}}
+    >>> A = SDM(elemsdict, (2, 2), QQ)
+    >>> A
+    {0: {1: 2}}
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:37]: The most useful thing to show here would be what this looks like as a Matrix. I don't think that the meaning of the dict of dicts structure is obvious without being explained.

File: sympy/polys/matrices/sdm.py [hunk 7]
```diff
@@ -17,6 +17,43 @@ class SDM(dict):
...
+    {0: {0: 3, 1: 2}, 1: {0: 1, 1: 4}}
+
+    Multiplication
+
+    >>> A = SDM({0:{1: ZZ(2)}, 1:{0:ZZ(1)}}, (2, 2), ZZ)
+    >>> B = SDM({0:{0: ZZ(3)}, 1:{1:ZZ(4)}}, (2, 2), ZZ)
+    >>> A*B
+    {0: {1: 8}, 1: {0: 3}}
+    >>> A*ZZ(2)
+    {0: {1: 4}, 1: {0: 2}}
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:55]: These examples can be condensed. We don't need to create new matrices for each demonstration.

File: sympy/polys/matrices/sdm.py [hunk 8]
```diff
@@ -45,14 +82,82 @@ def __repr__(self):
 
     @classmethod
     def new(cls, sdm, shape, domain):
+        """
+
+        Parameters
+        ==========
+        sdm: A dict of dicts for non-zero elements in SDM
+        shape: Represents dimension of SDM
+        domain: Represents :py:class:`~.Domain` of SDM
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:91]: There should be a blank line after `=====`.

The description of the parameters should actually say what the are e.g. `shape` is a tuple.

File: sympy/polys/matrices/sdm.py [hunk 9]
```diff
@@ -45,14 +82,82 @@ def __repr__(self):
...
+        =======
+
+        An :py:class:`~.SDM` object
+
+        Examples
+        ========
+
+        >>> from sympy.polys.matrices.sdm import SDM
+        >>> from sympy import QQ
+        >>> elemsdict = {0:{1:2}, 1:{}}
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:103]: The empty dict shouldn't be there. The `2` should be `QQ(2)`

File: sympy/polys/matrices/sdm.py [hunk 10]
```diff
@@ -75,28 +211,108 @@ def to_list(M):
...
+        belonging to the specified domain
+
+        Examples
+        ========
+
+        >>> from sympy.polys.matrices.sdm import SDM
+        >>> from sympy import QQ
+        >>> A = SDM.zeros((2, 3), QQ)
+        >>> A
+        {}
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:246]: This would be easier to understand if you also show what the corresponding matrix would be.

File: sympy/polys/matrices/sdm.py [hunk 11]
```diff
@@ -75,28 +211,108 @@ def to_list(M):
...
+        {0: {0: 1}, 1: {1: 1}}
+
+        """
         one = domain.one
         sdm = {i: {i: one} for i in range(size)}
         return cls(sdm, (size, size), domain)
 
     def transpose(M):
-        MT = sdm_transpose(M)
+        """
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:273]: This method should still use `sdm_transpose`

File: sympy/polys/matrices/sdm.py [hunk 12]
```diff
@@ -308,7 +751,7 @@ def sdm_irref(A):
     >>> pivots
     [0, 1]
 
-    The analogous calculation with :py:class:`~.Matrix` would be
+    The analogous calculation with Matrix class would be
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:754]: Why is this changed?

**Comment by ks147** [comment]
@oscarbenjamin please review

File: sympy/polys/matrices/sdm.py [hunk 13]
```diff
@@ -13,10 +13,52 @@
...
+                0 & \frac{1}{2} \\
+                0 & 0 \end{array} \right]
+
+
+    >>> from sympy.polys.matrices.sdm import SDM
+    >>> from sympy import QQ
+    >>> elemsdict = {0:{1:QQ(1, 2)}}
+    >>> A = SDM(elemsdict, (2, 2), QQ)
+    >>> A
+    {0: {1: 1/2}}
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:44]: It would be good so short the result of `A.to_Matrix()` here.

File: sympy/polys/matrices/sdm.py (hunk 10 above)
Comment (oscarbenjamin) [reply sympy/polys/matrices/sdm.py:256]: You could show what `A.to_Matrix()` gives here

File: sympy/polys/matrices/sdm.py [hunk 14]
```diff
@@ -75,28 +217,105 @@ def to_list(M):
...
+        >>> from sympy.polys.matrices.sdm import SDM
+        >>> from sympy import QQ
+        >>> A = SDM({0:{1:QQ(2)}, 1:{}}, (2, 2), QQ)
+        >>> A.transpose()
+        {1: {0: 2}}
+
+        """
         MT = sdm_transpose(M)
-        return M.new(MT, M.shape[::-1], M.domain)
+        return MT
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:298]: Why is this changed? Does this return a dict or `SDM`?

File: sympy/polys/matrices/sdm.py [hunk 15]
```diff
@@ -193,6 +614,31 @@ def hstack(A, *B):
...
+
+        We can create a polynomial using the
+        coefficients using :py:class:`~.Poly`
+
+        >>> from sympy import Symbol
+        >>> from sympy.polys import Poly
+        >>> x = Symbol('x')
+        >>> p = Poly(A.charpoly(), x, domain=A.domain)
+        >>> p
+        Poly(x**2 - 5*x - 2, x, domain='QQ')
```
Comment (oscarbenjamin) [inline sympy/polys/matrices/sdm.py:639]: This is very verbose. You could just combine the Symbol and Poly imports with QQ above and then this is just one line:
```
>>> Poly(A.charpoly(), Symbol('x'), domain=A.domain)
```

Comment (ks147) [reply sympy/polys/matrices/sdm.py:639]: Ok

File: sympy/polys/matrices/sdm.py (hunk 13 above)
Comment (ks147) [reply sympy/polys/matrices/sdm.py:44]: I don't think SDM class has `to_Matrix()`,  `DomainMatrix` on the other hand does have that functionality

**Comment by oscarbenjamin** [comment]
Okay, looks good.
